### DISCUSSION TITLE
User Management & Password hashing

### DISCUSSION THREAD

**Discussion by madsem** [filamentphp/filament#2138] [Help] [answered]
Hi,

I created a resource for my User model, when saving a User I realized that passwords are saved in plain text.
Now I'm looking for a way how I can hash them, I saw the Lifecycle Hooks, but it seems you don't have access to the actual field value in there?

How would I hash passwords when creating a User, and when editing a User?

**Comment by ryangjchandler** [comment, marked as answer]
@madsem You can use `dehydrateUsing()`, provide a callback and return a hashed version of the password. https://filamentphp.com/docs/2.x/forms/advanced#dehydration

**Reply by danharrin** [reply]
Alternatively, use `mutateFormDataBeforeCreate()`, which is a dedicated hook for pages.

https://filamentphp.com/docs/2.x/admin/resources#customizing-data-before-saving

**Reply by Polfo** [reply]
New link: https://filamentphp.com/docs/2.x/admin/resources/creating-records#customizing-data-before-saving

**Comment by madsem** [comment]
Thanks @ryangjchandler  & @danharrin this worked like a charm.

Is it possible to also have the password field be `optional` on `edit` pages, so a user doesn't always have to fill it in again when editing another field?

**Reply by danharrin** [reply]
Certainly - 

```php
->required(fn ($livewire) => $livewire instanceof CreateRecord) // the field is only required on the Create page
->dehydrated(fn ($state) => filled($state)) // the field is only saved to the DB when it has a value ($state)
```

**Reply by madsem** [reply]
noice, thanks so much @danharrin. Appreciate you guys help getting me going here :)

**Reply by alpondith** [reply]
`dehydrated(true/false)` decides the field will be present in the array or not to save the data.
`dehydrateStateUsing(..........)` decides how it should be modified before saving.

i faced the same problem then i came up with this solution. 

>                 TextInput::make('password')
>                        ->password()
>                        ->dehydrated(fn($state) => filled($state))
>                        ->dehydrateStateUsing(fn($state) => Hash::make($state))
>                        ->maxLength(255),

**Reply by danharrin** [reply]
It's all written down here - https://filamentphp.com/tricks/password-form-fields

**Comment by alpondith** [comment]
`dehydrated(true/false)` decides the field will be present in the array or not to save the data.
`dehydrateStateUsing(..........)` decides how it should be modified before saving.

i faced the same problem then i came up with this solution. 

>                 TextInput::make('password')
>                        ->password()
>                        ->dehydrated(fn($state) => filled($state))
>                        ->dehydrateStateUsing(fn($state) => Hash::make($state))
>                        ->maxLength(255),
